### ISSUE TITLE
As a user, if I specify minute and seconds as part of the reference_dates period, I want those minutes and seconds to passed through to web services used as inputs

### ISSUE THREAD

**Issue by HankHerr-NOAA** [NOAA-OWP/wres#292]
See #62. As part of testing differences, we noticed that the WRES was unable to parse this in a WRDS response (the date uses a "basic" ISO8601 format):

```json
            "reference_time": "20240809T000000Z",
```

That date/time string resulted from a request posted manually to WRDS that included minutes and seconds. In examining that failure, we realized that the WRDS reader was not respecting the minutes and seconds when provided as part of the `reference_dates` declaration:

```yaml
reference_dates:
  minimum: '2024-08-09T00:00:00Z'
  maximum: '2024-08-19T15:14:29Z'
```

This ticket can be resolved when:

- We modify the WRES to pass through the minutes and seconds of a reference dates declaration to web service data calls, including both WRDS (which can read it as #62 confirms) and USGS NWIS (if it doesn't already).
- We modify the WRES to be able to successfully read the resulting JSON returned by the web service. This will likely require a custom date parser, since the current parser cannot handle the ISO8601 basic format shown above. 

Thanks,

Hank

**Comment by james-d-brown** [comment]
In particular, note: https://github.com/NOAA-OWP/wres/issues/62#issuecomment-2314847608

**Comment by james-d-brown** [comment]
I see there is already a custom `NwmDataPointDeserializer`, but that isn't used for `NwmForecast`, rather `NwmDataPoint`. Probably need to have a custom deserializer with more limited scope for the datetime, leveraged by both classes.

**Comment by james-d-brown** [comment]
This appears to be the magic sauce for lenient parsing of "basic" ISO8601 datetime strings in which the minutes and/or seconds are optional:

```
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern( "yyyyMMdd'T'HH[mm[ss]]'Z'" )
                                                       .withZone( ZoneId.of( "UTC" ) );
```

**Comment by james-d-brown** [comment]
Progress in https://github.com/NOAA-OWP/wres/tree/issue292, which deals with the response. Will work on the request side next week.

**Comment by james-d-brown** [comment]
This is the special snowflake inside the `WrdsNwmReader` that formats datetimes to ISO8601 "basic" strings with hours only:

```
    /**
     * The WRDS NWM API uses a concise ISO-8601 format rather than RFC3339 instant.
     * @param instant the instance
     * @return the ISO-8601 instant string
     */
    private static String iso8601TruncatedToHourFromInstant( Instant instant )
    {
        String dateFormat = "uuuuMMdd'T'HHX";
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern( dateFormat )
                                                       .withZone( ZoneId.of( "UTC" ) );
        return formatter.format( instant );
    }
```

Need to fix that, but first need a (failing) unit test that expects minutes and seconds in the (mock) server request.

**Comment by james-d-brown** [comment]
Have a failing unit test. Now to make it pass.

**Comment by james-d-brown** [comment]
Have the test passing, but I noticed something odd about the assumed WRDS NWM api that may not actually make any difference when serving forecast data, but our reader assumes that the valid time query parameter is camel case, `validTime`. This can't be right, given that the reference time parameter is snake case, `valid_time`.

**Hank**, when you have a minute, could you check the WRDS NWM API query parameters. My expectation is that the api expects `vaid_time` for consistency with `reference_time` and our use of `validTime=all` is effectively being ignored, but probably defaulting to the same assumption.

**Comment by james-d-brown** [comment]
Pushed an update that respects the minutes and seconds portion of a declaration in the requests to WRDS NWM. This ticket should be done, but I await Hank's input on the above before merging into main.

**Comment by james-d-brown** [comment]
OK, thanks, will remove all references to it.

**Comment by HankHerr-NOAA** [comment]
James,

I believe I'll be UATing this one with the next deployment, and you may not be able to provide details then. I think the test, from my end, is to specify the reference time period using minutes and seconds. That should trigger the WRES to include those minutes and seconds when calling WRDS, which should then result WRDS returning reference times that include minutes and seconds, which the WRES should now be able to able to parse.  

Does that sound right? 

If I can, I'll try to post a test of it using a standalone just to make sure it works by Thursday. 

Hank

**Comment by HankHerr-NOAA** [comment]
Hmmm... What I just described is pretty much every evaluation we perform using a declared `reference_dates`.  So, I just need to test that WRES can still read the WRDS response after this ticket's change and get identical results before and after.  Right?

Hank

**Comment by james-d-brown** [comment]
Testing the response is easier because you can load a response from file with datetimes that include minutes/seconds, as in the OP. This produced an error before. Testing the request is a bit trickier, but also less important. I don't think any use of minutes/seconds will actually discriminate different data from wrds, for example. However, this is also less important because it is unit tested. Running in debug mode for the relevant package will produce information about the request, so there's that.

Also bear in mind that we chunk by week ranges for WRDS NWM, so no request for more than a week of data will use precisely the declared datetimes, although the last chunk will end with the same minuteseconds as the declared end datetime.

**Comment by james-d-brown** [comment]
Oh, and one more thing, the reader starts the week ranges on the previous Sunday at the declared datetime with minutes and seconds zeroed, which was a choice we made to increase re-use of data, so that nuance is a "feature" not a bug.

**Comment by HankHerr-NOAA** [comment]
I tested the requests using a standalone with debug output turned on. I could see the minutes and seconds getting into the requests for both WRDS and USGS NWIS.

Let me setup a test for handling the response by using what triggered the error I was seeing. Hopefully, #62 has the declaration I need,

Hank

**Comment by james-d-brown** [comment]
The request part of that is going to get backed out, because it was a mistake and is unneeded. We request by weekly blocks to promote re-use, so it doesn't matter what a user asks for, providing our requests encapsulate it. The response is the only thing to test here.

**Comment by james-d-brown** [comment]
See #309.

**Comment by james-d-brown** [comment]
( And note that the response problem will never get hit in reality, only when constructing a manual request to wrds and then attempting to read that response. )

**Comment by HankHerr-NOAA** [comment]
Oh, got it. Okay.

Yeah, I remembered I can trigger the right kind of response from WRDS without manually obtaining the data. I'm just going to pull down the data from when we first encountered the problem in #62 and confirm that we no loner encounter the problem.

Hank

**Comment by james-d-brown** [comment]
Yes, you basically need to construct a response that has minutes/seconds in the forecast dates. Bear in mind that the first time Arvin uploaded data to test this, the period was wrong, so make sure you grab the correct files from that ticket else create yourself.
